File: Topcoder/SRM 592/LittleElephantAndPermutationDiv1.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>

#define MOD 1000000007

using namespace std;

class LittleElephantAndPermutationDiv1 {
public:
    int dp[51][51][2501];
    bool seen[51][51][2501];
    int n, k;
    
    int solve(int x, int fx, int sum) {
        if (x == n+1) {
            if (fx == 0 && sum >= k)
                return 1;
            else
                return 0;
        }   

        if (seen[x][fx][sum])
            return dp[x][fx][sum];

        seen[x][fx][sum] = true;

        int &ret = dp[x][fx][sum];
        ret = 0;

        ret = (ret + solve(x+1, fx+1, sum))%MOD;
        ret = (ret + solve(x+1, fx, sum+x))%MOD;
        if (fx > 0)
            ret = (ret + solve(x+1, fx-1, sum + 2*x)*1LL*fx*fx%MOD)%MOD;
        ret = (ret + solve(x+1, fx, sum + x)*1LL*2*fx%MOD)%MOD;

        return ret; 
    }

    int getNumber(int _N, int _K) {
        memset(dp, 0, sizeof dp);
        memset(seen, false, sizeof seen);
        n = _N; k = _K;
        int ans = solve(1, 0, 0);

        for(int i = 1; i <= n; i++)
            ans = ans*1LL*i%MOD;

        return ans;
    }
};

// CUT begin
template <typename T> string pretty_print(T t) { stringstream s; typeid(T) == typeid(string) ? s << "\"" << t << "\"" : s << t; return s.str(); }

bool do_test(int N, int K, int __expected, int caseNo) {
    cout << "  Testcase #" << caseNo << " ... ";

    time_t startClock = clock();
    LittleElephantAndPermutationDiv1 *instance = new LittleElephantAndPermutationDiv1();
    int __result = instance->getNumber(N, K);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << pretty_print(__expected) << endl;
        cout << "           Received: " << pretty_print(__result) << endl;
        return false;
    }
}

bool run_testcase(int __no) {
    switch (__no) {
        case 0: {
            int N = 1;
            int K = 1;
            int __expected = 1;
            return do_test(N, K, __expected, __no);
        }
        case 1: {
            int N = 2;
            int K = 1;
            int __expected = 4;
            return do_test(N, K, __expected, __no);
        }
        case 2: {
            int N = 3;
            int K = 8;
            int __expected = 18;
            return do_test(N, K, __expected, __no);
        }
        case 3: {
            int N = 10;
            int K = 74;
            int __expected = 484682624;
            return do_test(N, K, __expected, __no);
        }
        case 4: {
            int N = 50;
            int K = 1000;
            int __expected = 539792695;
            return do_test(N, K, __expected, __no);
        }

        // Your custom testcase goes here
        case 5:
            break;
        default: break;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed,ios::floatfield);
    cout.precision(2);
    cout << "LittleElephantAndPermutationDiv1 (500 Points)" << endl << endl;

    int nPassed = 0, nAll = 0;
    if (argc == 1)
        for (int i = 0; i < 5; ++i) nAll++, nPassed += run_testcase(i);
    else
        for (int i = 1; i < argc; ++i) nAll++, nPassed += run_testcase(atoi(argv[i]));
    cout << endl << "Passed : " << nPassed << "/" << nAll << " cases" << endl;

    int T = time(NULL) - 1380526799;
    double PT = T / 60.0, TT = 75.0;
    cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
    cout << "Score  : " << 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    return 0;
}
// CUT end

File: Topcoder/SRM 592/LittleElephantAndPermutationDiv1.java
public class LittleElephantAndPermutationDiv1 {
	public int getNumber(int N, int K) {
		return 0;
	}

// CUT begin
	static void runTestcase(int __case) {
		switch (__case) {
			// Your custom testcase goes here
			case -1:
				//doTest(N, K, expected, __case);
				break;

			case 0: {
				int N = 1;
				int K = 1;
				int __excepted = 1;
				doTest(N, K, __excepted, __case);
				break;
			}
			case 1: {
				int N = 2;
				int K = 1;
				int __excepted = 4;
				doTest(N, K, __excepted, __case);
				break;
			}
			case 2: {
				int N = 3;
				int K = 8;
				int __excepted = 18;
				doTest(N, K, __excepted, __case);
				break;
			}
			case 3: {
				int N = 10;
				int K = 74;
				int __excepted = 484682624;
				doTest(N, K, __excepted, __case);
				break;
			}
			case 4: {
				int N = 50;
				int K = 1000;
				int __excepted = 539792695;
				doTest(N, K, __excepted, __case);
				break;
			}
			default: break;
		}
	}

	static void doTest(int N, int K, int __expected, int caseNo) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		LittleElephantAndPermutationDiv1 instance = new LittleElephantAndPermutationDiv1();
		int __result = 0;
		try {
			__result = instance.getNumber(N, K);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		nAll++;
		System.err.print(String.format("  Testcase #%d ... ", caseNo));

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			nPassed++;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
		}
	}

	static int nExample = 5;
	static int nAll = 0, nPassed = 0;

	public static void main(String[] args){
		System.err.println("LittleElephantAndPermutationDiv1 (500 Points)");
		System.err.println();
		if (args.length == 0)
			for (int i = 0; i < nExample; ++i) runTestcase(i);
		else
			for (int i = 0; i < args.length; ++i) runTestcase(Integer.parseInt(args[i]));
		System.err.println(String.format("%nPassed : %d/%d cases", nPassed, nAll));

		int T = (int)(System.currentTimeMillis() / 1000) - 1380526799;
		double PT = T / 60.0, TT = 75.0;
		System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}
// CUT end
}
